go.mod

```diff
@@ -472,6 +476,7 @@ replace (
 	github.com/xordataexchange/crypt => github.com/xordataexchange/crypt v0.0.3-0.20170626215501-b2862e3d0a77
 	go.etcd.io/bbolt => go.etcd.io/bbolt v1.3.3
 	go.etcd.io/etcd => go.etcd.io/etcd v0.0.0-20191023171146-3cf2f69b5738
+	go.mongodb.org/mongo-driver => go.mongodb.org/mongo-driver v1.1.2
```

Comment:
what pulled this in?

Reply:
Only called pin-dependencies and update-vendor. But interesting indeed.

Reply:
`go-openapi/strfmt`.  It got support for Mongo's bson format. 

We should immediately start restricting the supported formats to what we have in 1.16. I fear we get more and more format without noticing.

Reply:
Agree. This is concerning, both from what it exposes in our API, and what it pulls in as dependencies. I think we need to open upstream issues requesting things like this be split so all downstreams don't pick up this dependency, or look into replacing this library long term.

Reply:
> We should immediately start restricting the supported formats to what we have in 1.16.

do we restrict format today? or is it allowed to be arbitrary and some special values just get magical validation behavior from go-openapi?

even if we lock down our API surface, I don't want to pick up dependencies like this, so I don't think we should take this unless we have a plan for dropping it

Reply:
Agreed.  We need a very tight control of public APIs, and do not want to be dependent on dependencies for api stability.

Reply:
https://github.com/kubernetes/kubernetes/pull/85223 as a first step.

Reply:
> do we restrict format today? or is it allowed to be arbitrary and some special values just get magical validation behavior from go-openapi?

By definition formats are best effort to be checked by OpenAPI implementors. We allow any format string. go-openapi checks two handful of format strings, see PR.

Reply:
#85223 looks good for controlling our API surface.

Reply:
@sttts pointed out that our openapi validation already required a mongo driver to do bsonobjectid validation (github.com/globalsign/mgo). this drops that unofficial mongo code (7771 lines worth) in favor of the official driver (13402 lines worth).

I'm ok with the following plan:
1. control api surface in #85223
2. merge this to fix the items validation bug
3. open issue upstream in openapi-validator/strfmt to see if they'll refactor to drop the mongo dep
4. open an issue in k/k consider ways to clean up and control our openapi validator dependency, even if it means extracting the narrow pieces we depend on

Reply:
github.com/globalsign/mgo replaced by go.mongodb.org/mongo-driver, the former unmaintained according to the go-openapi project.